EIP1559

Spacesai1or opened the pull request:
Uses [ethers.js' implementation/defaults reasoning](https://github.com/ethers-io/ethers.js/blob/c5bca7767e3f3d43e3d0bd3c9e9420321ee9907a/packages/abstract-signer/src.ts/index.ts#L199) as inspiration

closes #4150

~**NOTE** Extra file changes are a result of branch off #4118, since it contains the needed EIP-1559 changes~

Spacesai1or committed:
npm run build for 1.4.0-rc.0

Spacesai1or committed:
v1.4.0-rc.0

Spacesai1or committed:
Update scripts/e2e.geth.instamine.sh to use v1.10.3 of Geth

Spacesai1or committed:
Update scripts/e2e.geth.instamine.sh to use v1.10.3 of Geth

Spacesai1or committed:
Fix typo

Spacesai1or committed:
Fix merge conflict

Spacesai1or committed:
WIP

Spacesai1or committed:
Merge branch '1.x' into wyatt/eip1559

kchia commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -287,6 +320,30 @@ function _validateTransactionForSigning(tx) {
+                }
+            })
+        } catch (error) {
+            reject(e)
```
Looks like a typo here. Should say: ```reject(error)```

GregTheGreek replied:
```suggestion
            reject(error)
```

Spacesai1or replied:
Thank y'all!

ricmoo reviewed (commented):
Keep in mind that EIP-1559 transactions also have an accessList, so [this section](https://github.com/ChainSafe/web3.js/pull/4155/files#diff-cf46f2521b1de659430e09e654742a41c3696dbdddc3cd563ca4e2c02d18ffacR172) might need some re-ordering.

There are also lots of decisions to make about incompatible properties (e.g. `maxFeePerGas` and `gasPrice` specifies).

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -162,6 +172,9 @@ Accounts.prototype.signTransaction = function signTransaction(tx, privateKey, ca
             if (transaction.accessList) {
                 // EIP-2930
                 transaction.type = "0x01"
+            } else if (transaction.maxPriorityFeePerGas || transaction.maxFeePerGas) {
```
I think this should be the first case, since in practice 1559 can be appended to any transaction type.

ricmoo replied:
<strike>But what happens if they specify `accessList` and `maxFeePerGas`?</strike>

Sorry, misread. Yes, I agree. That’s what I meant by re-ordering. :)

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -241,21 +254,41 @@ Accounts.prototype.signTransaction = function signTransaction(tx, privateKey, ca
+                chainId: args[0],
+                nonce: args[2],
+                networkId: args[3],
+                ...args[1] // Will either be gasPrice or maxFeePerGas and maxPriorityFeePerGas
```
ocd killing me on this, can you push `_handleTxPricing` to the end of `Promise.all(..)` so it inherits the 3rd index? Or just re-order this

GregTheGreek reviewed (changes requested):
First glance, need to take a more thorough look

Commit:
Fix type in error catch

Co-authored-by: Gregory Markou <16929357+GregTheGreek@users.noreply.github.com>

Spacesai1or commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -241,21 +254,41 @@ Accounts.prototype.signTransaction = function signTransaction(tx, privateKey, ca
     return Promise.all([
         isNot(tx.chainId) ? _this._ethereumCall.getChainId() : tx.chainId,
-        isNot(tx.gasPrice) ? _this._ethereumCall.getGasPrice() : tx.gasPrice,
+        // isNot(tx.gasPrice) ? _this._ethereumCall.getGasPrice() : tx.gasPrice,
```
Remove commented code

Spacesai1or committed:
Remove commented code

Spacesai1or committed:
Update index for tx pricing info

Spacesai1or committed:
Correct order of if statements to properly assign tx type

Spacesai1or committed:
Update maxFeePerGas calculation

Spacesai1or committed:
Init _handleTxType function

Spacesai1or committed:
Update _handleTxPricing to use gasPrice for eip-1559 txs

Spacesai1or commented:
Thank you @kchia , @ricmoo , and @GregTheGreek for taking a look at this

I've
- added a `_handleTxType` method that includes property checking as @ricmoo mentioned (copies [checks from Ethers](https://github.com/ethers-io/ethers.js/blob/2a7ce0e72a1e0c9469e10392b0329e75e341cf18/packages/abstract-signer/src.ts/index.ts#L215))
- updated `maxFeePerGas` calculation as per [Ethers](https://github.com/ethers-io/ethers.js/blob/ba6854bdd5a912fe873d5da494cb5c62c190adde/packages/abstract-provider/src.ts/index.ts#L247)
- updated `_handleTxPricing` to use `gasPrice` for `maxPriorityFeePerGas` and `maxFeePerGas` if network supports EIP-1559, but user provided `gasPrice` as per [Ethers](https://github.com/ethers-io/ethers.js/blob/2a7ce0e72a1e0c9469e10392b0329e75e341cf18/packages/abstract-signer/src.ts/index.ts#L246)

Thank you @ricmoo for your R&D :)

Spacesai1or committed:
Fix bugs in _handleTxPricing

Spacesai1or committed:
Update tests for eip-1559 changes

Spacesai1or committed:
Fix BN bug: .mul only accepts BN instances

Spacesai1or committed:
eth.accounts.signTransaction: Add London tests

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -81,7 +81,17 @@ var Accounts = function Accounts() {
+            }, function() {
+                return false
+            }]
+        }),
```
is this a new method?

Spacesai1or replied:
Sorta, this package doesn't make use of the RPC wrappers found in `web3-eth`, instead needed methods are redefined in the package. I added this method to be able to get a block to find the `baseFeePerGas`

AndreMiras replied:
You mean we can't leverage the [getBlock](https://github.com/ChainSafe/web3.js/blob/v1.4.0/packages/web3-eth/src/index.js#L420-L426) (also refs [blockCall](https://github.com/ChainSafe/web3.js/blob/v1.4.0/packages/web3-eth/src/index.js#L44-L46)) because it's defined in the `web3-eth` package and not accessible from `web3-eth-accounts`?
In this case don't we have a dependency or at least a DRY issue that could be addressed somehow?

Spacesai1or replied:
@AndreMiras Yes, and we plan to address the problematic architecture in `4.x` (our rewrite), so for now it's just about making the least amount of change to `1.x` as possible, to refrain from introducing additional bugs

GregTheGreek commented on test/eth.accounts.signTransaction.js:
```diff
@@ -500,6 +557,52 @@ var tests = [
+        oldSignature: "0x02f8ca0180843b9aca00843b9aca0e826a4094f0109fc8df283027b6285cc889f5aa624eac1f55843b9aca0080f85bf859940000000000000000000000000000000000000101f842a00000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000060a780a0e3a2e10c7d3af3407ec2d38c64788d6673926e9b28d6d2e7df3c94cdf0548233a00ad3e5faafaf3a9350ab16c1be01 ...
+        transactionHash: "0xbc2c9edab3d4e3a795fa402b52d6149e874de15f0cc6c0858eb34e1fe1ef31fe",
+        messageHash: "0xa3a2cdc45e9cefb9a614ead90ce65f68bcf8a90dbe0ccbd84c1b62403bd05346"
+    },
```
Perhaps we can add a few tests:
- 1559 without `accessList`
- 1559 without the gas values

Spacesai1or replied:
A test without `gasPrice` or fee related properties exists [here](https://github.com/ChainSafe/web3.js/blob/4ebd78a771a046dd287ecf1cfdedc37986162a71/test/eth.accounts.signTransaction.js#L585)

I'll add a test that excludes `accessList`, and tests that verify the `maxFeePerGas` calculation needs to be added as well

Spacesai1or replied:
Added test [without](https://github.com/ChainSafe/web3.js/pull/4155/commits/43fc7aa08ea128a000f6be1d425170b62c62dac1) AccessList

Spacesai1or committed:
Fix merge conflicts

Spacesai1or committed:
Add EIP-1559 test without AccessList

Spacesai1or committed:
Add tx.common.hardfork check to _handleTxType

Spacesai1or committed:
Bug fixes

Spacesai1or commented:
I created a [manual test repo](https://github.com/spacesailor24/web3-eip1559-test) for this, since existing `web3-eth-accounts` tests don't allow for testing internal methods, don't give access to constructed transaction object, and I wanted to test with a live network

Spacesai1or committed:
Add additional undefined check in _handleTxType

Spacesai1or committed:
Add additional check for tx.hardfork in _handleTxType

Spacesai1or committed:
Add additional check for tx.hardfork in _handleTxType

Spacesai1or committed:
Update CHANGELOG

liamaharon commented:
Hi there, I get an error calling `web3.eth.getTranasction` with a type 2 transaction hash using a web3.js WS provider.

The error occurs here https://github.com/ChainSafe/web3.js/blob/1.x/packages/web3-core-helpers/src/formatters.js#L242, it looks like this function also needs updating for web3.js to be compatible with new transaction types introduced with London. I'm using v1.4.0.

Is support for type 2 transactions planned to be added in this PR?

Thanks!

Spacesai1or commented:
@liamaharon Hi there! Thank you bringing this to our attention. This PR is looking to add support for type 2 transactions (EIP1559), but it does look like we still need to update the output formatter code you provided to account for the additional fields

corymsmith commented:
> @liamaharon Hi there! Thank you bringing this to our attention. This PR is looking to add support for type 2 transactions (EIP1559), but it does look like we still need to update the output formatter code you provided to account for the additional fields

@spacesailor24 Is someone actively implementing those updates at the moment? If not I could look into that right now.

corymsmith commented:
@spacesailor24 I created this PR with this one as the base branch: https://github.com/ChainSafe/web3.js/pull/4167

Spacesai1or commented:
> @spacesailor24 I created this PR with this one as the base branch: https://github.com/ChainSafe/web3.js/pull/4167

Thank you @corymsmith! I'll give this a more thorough ￼review, as well as test it against my test repo today, but at first glance it looks good!

jdevcs commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -287,6 +313,93 @@ function _validateTransactionForSigning(tx) {
+                            maxFeePerGas = tx.gasPrice;
+                            delete tx.gasPrice;
+                        } else {
+                            maxPriorityFeePerGas = tx.maxPriorityFeePerGas || '0x3B9ACA00'; // 1 Gwei
```
Its not part of the EIP ( specs )  to must have default 'priority fee per gas' 1 Gwei. Its only suggested to set this default.

so I suggest It should always be coming from web3 user only, or if we are setting default then must notify / inform user as well.

GregTheGreek replied:
This was something that was suggested on a call with wallet/library implementers, but we should make a note in the docs

Spacesai1or replied:
We'll add `maxPriorityFeePerGas` being defaulted to `1 Gwei` to the docs, but I think the understanding was that a tx without a priority fee wouldn't get processed/be excluded from the mem pool

jdevcs commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -287,6 +313,93 @@ function _validateTransactionForSigning(tx) {
+                ]).then(responses => {
+                    const [block, gasPrice] = responses;
+                    if (
+                        (tx.type !== '0x0' && 
```
Its already checked https://github.com/ChainSafe/web3.js/pull/4155/files#diff-cf46f2521b1de659430e09e654742a41c3696dbdddc3cd563ca4e2c02d18ffacR352 for legacy Tx types. May be we can remove this double check or is there any use of additional check?

Spacesai1or replied:
So while this is double checking `tx.type`, it's for different reasons:

[Here](https://github.com/ChainSafe/web3.js/blob/572e416e197156032b33a767e1457933afc2a68e/packages/web3-eth-accounts/src/index.js#L357) we are short circuiting and returning `gasPrice` if it's apart of the `tx` object and `tx.type` is `0x0` or `0x1` since these txs will not have the EIP1559 fields and will be using the user provided `gasPrice`

And [here](https://github.com/ChainSafe/web3.js/blob/572e416e197156032b33a767e1457933afc2a68e/packages/web3-eth-accounts/src/index.js#L365) we are verifying we aren't of type `0x0` or `0x1` before we add EIP1559 fields

I admit that this code is a bit convoluted, but it's essentially just a condensed version of Ether's EIP1559 implementation

corymsmith committed:
Handling EIP1559 transactions in outputTransactionFormatter (#4167)

* Handing a null gasPrice and supporting maxFeePerGas and maxPriorityFeePerGas

* Updating gasPrice check

Spacesai1or commented:
What remains to be done before this is ready for final review:

- [x] #4150 
- [x] Implement [these](https://github.com/spacesailor24/web3-eip1559-test/blob/main/index.js) tests within the repo using `geth-dev-assistant`

Spacesai1or committed:
Revert geth docker version tag

Spacesai1or committed:
Add additional EIP-2930 and EIP-1559 tests

jdevcs commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -287,6 +313,93 @@ function _validateTransactionForSigning(tx) {
+                            maxFeePerGas = tx.maxFeePerGas ||
+                                utils.toHex(
+                                    utils.toBN(block.baseFeePerGas)
+                                        .mul(utils.toBN(2))
```
Also will be imp to highlight that web3 is doing it and why "baseFeePerGas **times 2**" to web3 users in doc.

Commit:
Merge branch '1.x' into wyatt/eip1559

Spacesai1or committed:
Merge branch 'wyatt/eip1559' of github.com:ChainSafe/web3.js into wyatt/eip1559

kchia commented on test/eth.accounts.signTransaction.js:
```diff
@@ -24,6 +24,63 @@ var commonBerlin = {
     hardfork: 'berlin'
 };
 
+var commonLondon = {
```
I'd recommend using `let` instead of `var` to declare variables. `let` is a more modern syntax than `var` and it aims to resolve some weird hoisting behaviors that `var` has that could lead to bugs.

Spacesai1or replied:
Hey @kchia, thank you for the input! Unfortunately most of Web3.js was written a while back, before modern JavaScript techniques were widely available/supported. We're currently working on a `4.x` version this uses TypeScript and modern JS features all the way up to ES2020. So for `1.x` and `3.x` the concern is to limit the amount of changes made to reduce introducing any new bugs

corymsmith commented:
Any updates on this @spacesailor24, need any help?

Spacesai1or commented:
> Any updates on this @spacesailor24, need any help?

Hey there! Thanks for the offer, looks like we just got the [failing tests to pass](https://github.com/ChainSafe/web3.js/pull/4171). The team is going to have a sync call around this tomorrow as a sanity check, then this PR will most likely get merged and released as an RC tomorrow

corymsmith commented:
> > Any updates on this @spacesailor24, need any help?
> 
> Hey there! Thanks for the offer, looks like we just got the [failing tests to pass](https://github.com/ChainSafe/web3.js/pull/4171). The team is going to have a sync call around this tomorrow as a sanity check, then this PR will most likely get merged and released as an RC tomorrow

Awesome, once its released I can test it out in our staging environment, thanks for the update!

Commit:
EIP 1559 Debug #2 (#4171)

* Add additional EIP-2930 and EIP-1559 tests

* Update gasPrice to geth default minimum

* Replace hex number with number string

* Add test skip if using Ganache

* Add done function to ganache skip

* Add done function to ganache skip

* Update gasPrice params

* Bump gasPrice to Geth default

* Bump docker geth version to stable

* Gas price type number for 1 Gwei

* gasPrice type and backfills timeout change

* gasPrice type and test acct creation

* added nonce for geth auto nonce issue, removed auto acct creation

* Remove redundant London test. Add gasLimit field to London test

* Add validation for tx.maxPriorityFeePerGas and tx.maxFeePerGas

* Add additional 1559 tests

Co-authored-by: jdevcs <junaid@chainsafe.com>
Co-authored-by: jdevcs <junaid@chainsafe.io>

Spacesai1or committed:
Mergin conflicts

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
-        tx.chainId < 0) {
+    if (tx.gas && tx.gasPrice) {
+        if (tx.gas < 0 || tx.gasPrice < 0) {
+            return new Error('Gas, gasPrice, nonce or chainId is lower than 0');
```
```suggestion
            return new Error('Gas or gasPrice is lower than 0');
```

Spacesai1or replied:
I wanted to change these, but technically it's a breaking change since we'll be returning a different error message to the user

GregTheGreek replied:
It's not breaking the api or changing the behaviour, its still returning an error, so it would be caught with a `.catch()`. In general its not a great practice to be filtering errors based on exact matches...

I think in this case we would consider this to be a bug, since its now a misleading error message. Just my two cents.

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
+    }
+
+    if (tx.nonce < 0 || tx.chainId < 0) {
         return new Error('Gas, gasPrice, nonce or chainId is lower than 0');
```
```suggestion
        return new Error('Nonce or chainId is lower than 0');
```

GregTheGreek reviewed (changes requested):
Partial review.

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
+    const hasEip1559 = (tx.maxFeePerGas !== undefined || tx.maxPriorityFeePerGas !== undefined);
+    if (tx.gasPrice !== undefined && (tx.type === '0x2' || hasEip1559))
+        throw Error("eip-1559 transactions don't support gasPrice");
+    if ((tx.type === '0x1' || tx.type === '0x0') && hasEip1559)
```
here you use `tx.type` but everywhere else you use `txType` from line 325

Spacesai1or replied:
Updated

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
+    } else if (
+        tx.accessList ||
+        (
+            (tx.common && tx.common.hardfork && tx.common.hardfork.toLowerCase() === 'berlin') ||
```
The following is possible:
```
const {Hardfork} = require("@ethereumjs/common");

(tx.common && tx.common.hardfork && tx.common.hardfork === Hardfork.Berlin)
```

Spacesai1or replied:
Updated

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
+                (
+                    tx.type === undefined ||
+                    tx.type === '0x0' ||
+                    tx.type === '0x1'
```
this can be simplified to be 
```
tx.type < "0x2"
```

Spacesai1or replied:
Updated

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
+                    tx.type === undefined ||
+                    tx.type === '0x0' ||
+                    tx.type === '0x1'
+                ) && tx.gasPrice !== undefined)
```
This is checking that `tx.gasPrice` is equal to something, if its already set why are we setting it again?

Spacesai1or replied:
You're right that `tx.gasPrice` is already set, so returning the value here is redundant, but returning nothing will cause [this](https://github.com/ChainSafe/web3.js/pull/4155/checks?check_run_id=3119556049#step:6:420) error as it fails the check [here](https://github.com/ChainSafe/web3.js/blob/b3ab709f48fea959938c9d24845b5a7dd7bb1f1c/packages/web3-eth-accounts/src/index.js#L272). We could refactor the code to not call `_handleTxPricing` if `tx.gasPrice` is set and `tx.type < '0x2'`, but this doesn't really buy us anything, and moves some tx pricing logic out of `_handleTxPricing`. So I think it makes the most sense to keep how it is, and short circuit the `_handleTxPricing`

GregTheGreek commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,117 @@ function _validateTransactionForSigning(tx) {
+                    const [block, gasPrice] = responses;
+                    if (
+                        (tx.type !== '0x0' && 
+                        tx.type !== '0x1') &&
```
Simplified version?
```
if (tx.type == '0x2' || (block && block.baseFeePerGas)) { ... }
```

Spacesai1or replied:
Updated

Commit:
Update error message

Co-authored-by: Gregory Markou <16929357+GregTheGreek@users.noreply.github.com>

Commit:
Update error message

Co-authored-by: Gregory Markou <16929357+GregTheGreek@users.noreply.github.com>

Spacesai1or committed:
Merge branch 'wyatt/eip1559' of github.com:ChainSafe/web3.js into wyatt/eip1559

Goldselleruk commented on CHANGELOG.md:
```diff
@@ -392,7 +392,7 @@ Released with 1.0.0-beta.37 code base.
 ### Changed
 
-- Changed Geth Docker verision from `stable` to `1.10.3` in `e2e.geth.instamine.sh`
+- Changed Geth Docker verision from `stable` to `1.10.3` in `e2e.geth.instamine.sh` and `scripts/e2e.geth.instamine.sh` (#4154)
```
Gdzie gold

Goldselleruk reviewed (commented):
Gdzie gold

Goldselleruk reviewed (dismissed):
Gdzie gold

Spacesai1or committed:
Update use of tx.type

Spacesai1or committed:
Replace hardfork strings with enum

Spacesai1or committed:
Type check refactors for _handleTxPricing

Spacesai1or committed:
Resolve tx.gasPrice if set and tx.type < 0x2 in _handleTxPricing

Spacesai1or committed:
Fix _handleTxType logic

luu-alex commented on packages/web3-eth-accounts/src/index.js:
```diff
@@ -272,20 +297,120 @@ function _validateTransactionForSigning(tx) {
+function _handleTxPricing(_this, tx) {
+    return new Promise((resolve, reject) => {
+        try {
+            if (tx.type < '0x2' && tx.gasPrice !== undefined) {
```
when `tx.type = undefined` and is checked under `tx.type < "0x2"` it'll result into false
perhaps:
```
if (( tx.type === 'undefined' || tx.type) < '0x2' && ...
 ```

Spacesai1or replied:
It's intended to be false if `tx.type` is `undefined`. We only want to resolve the provided `tx.gasPrice` if `tx.type` is `0x0` or `0x1`. If `undefined`, we don't know if the network supports `EIP-1559`, and whether we should use `tx.gasPrice` or the 1559 properties

Spacesai1or commented:
@corymsmith [v1.5.0-rc.0](https://github.com/ChainSafe/web3.js/releases/tag/v1.5.0-rc.0) has been released

ricmoo commented:
FYI, if you need more test cases, here are the ethers generated (with EthereumJS) testcases: https://github.com/ethers-io/ethers.js/blob/master/packages/testcases/testcases/typed-transactions.json.gz
